manta-pay/src/crypto/poseidon/parameters_hardcoded_test/generate_parameters_grain_deterministic.sage

```diff
@@ -0,0 +1,347 @@
+            intersection = S.intersection(vector_subspace)
+            basis_vectors += intersection.basis()
+        IS = V.subspace(basis_vectors)
+
+        if IS.dimension() >= 1 and IS != V:
+            return [False, 2]
+        for j in range(1, i+1):
+            S_mat_mul = subspace_times_matrix(S, M^j, t)
+            if S == S_mat_mul:
+                print("S.basis():\n", S.basis())
+                return [False, 3]
+
+    return [True, 0]
+
+# Returns True if the matrix is considered secure, False otherwise
+def algorithm_2(M, NUM_CELLS):
+    t = NUM_CELLS
+    s = 1
+
+    V = VectorSpace(F, t)
+    trail = [None, None]
+    test_next = False
+    I = range(0, s)
+    I_powerset = list(sage.misc.misc.powerset(I))[1:]
+    for I_s in I_powerset:
+        test_next = False
+        new_basis = []
+        for l in I_s:
+            new_basis.append(V.basis()[l])
+        IS = V.subspace(new_basis)
+        for i in range(s, t):
+            new_basis.append(V.basis()[i])
+        full_iota_space = V.subspace(new_basis)
+        for l in I_s:
+            v = V.basis()[l]
+            while True:
+                delta = IS.dimension()
+                v = M * v
+                IS = V.subspace(IS.basis() + [v])
+                if IS.dimension() == t or IS.intersection(full_iota_space) != IS:
+                    test_next = True
+                    break
+                if IS.dimension() <= delta:
+                    break
+            if test_next == True:
+                break
+        if test_next == True:
+            continue
+        return [False, [IS, I_s]]
+
+    return [True, None]
+
+# Returns True if the matrix is considered secure, False otherwise
+def algorithm_3(M, NUM_CELLS):
+    t = NUM_CELLS
+    s = 1
+
+    V = VectorSpace(F, t)
+
+    l = 4*t
+    for r in range(2, l+1):
+        next_r = False
+        res_alg_2 = algorithm_2(M^r, t)
+        if res_alg_2[0] == False:
+            return [False, None]
+
+        # if res_alg_2[1] == None:
+        #     continue
+        # IS = res_alg_2[1][0]
+        # I_s = res_alg_2[1][1]
+        # for j in range(1, r):
+        #     IS = subspace_times_matrix(IS, M, t)
+        #     I_j = []
+        #     for i in range(0, s):
+        #         new_basis = []
+        #         for k in range(0, t):
+        #             if k != i:
+        #                 new_basis.append(V.basis()[k])
+        #         iota_space = V.subspace(new_basis)
+        #         if IS.intersection(iota_space) != iota_space:
+        #             single_iota_space = V.subspace([V.basis()[i]])
+        #             if IS.intersection(single_iota_space) == single_iota_space:
+        #                 I_j.append(i)
+        #             else:
+        #                 next_r = True
+        #                 break
+        #     if next_r == True:
+        #         break
+        # if next_r == True:
+        #     continue
+        # return [False, [IS, I_j, r]]
```

Comment:
Why is this commented out?

Reply:
They're also commented out in the source file:
https://extgit.iaik.tugraz.at/krypto/hadeshash/-/blob/master/code/generate_parameters_grain.sage

Reply:
Yeah but then we don't want it because we don't need it.